Add `-source-tree` and `-build-tree` options to ocamldep

MisterDA opened the pull request:
ocamldep generates the list of dependencies for a given ocaml target file. The new `-source-tree` option allows to find dependencies in an other location than just the current working directory. The new `-build-tree` option allows to write targets (build artifacts) in an other location than just the current working directory.

This is useful for out-of-source builds and cross-compilation.

ocamldep outputs a trailing directory separator if the given string doesn't contain one, and respects `-slash` and `-no-slash` options. No other processing is made on the string. It is for instance possible to give a Makefile variable instead of a path.

MisterDA committed:
Add a -build-tree option to ocamldep

ocamldep generates the list of dependencies of a given OCaml target
file. The new -build-tree option adds a facility to add a (separate)
prefix to the targets and the dependencies in its make output. This
allows make to write targets in an other location than just the
current working directory. Dependencies of the target that are also
build artifacts are also found in this location.

This is useful for out-of-source builds and cross-compilation.

ocamldep outputs a trailing directory separator if the given string
doesn't contain one, and respects -slash and -no-slash options. No
other processing is made on the string. It is for instance possible to
give a Makefile variable instead of a path.

nojb commented:
I will review this one, as it is related to #12203.

shindere commented:
Many thanks, @MisterDA.

For sake of completeness and transparency, I'll say that I did
apre-review of the code before the PR got opened and it looked okay to
me. I suggested that we could perhaps add a test to the testsuite.

nojb commented on driver/makedepend.ml:
```diff
@@ -197,25 +199,37 @@ let print_filename s =
 let print_dependencies target_files deps =
   let pos = ref 0 in
-  let print_on_same_line item =
+  let dir_sep = if Sys.os_type = "Unix" then '/'
+                else if !Clflags.force_slash then '/' else '\\' in
```
```suggestion
  let dir_sep = if Sys.unix || !Clflags.force_slash then '/' else '\\' in
```

nojb commented on driver/makedepend.ml:
```diff
@@ -197,25 +199,37 @@ let print_filename s =
+  in
+  let print_source_tree, source_tree_len = print_prefix !source_tree
+  and print_build_tree, _ = print_prefix !build_tree in
+  let print_on_same_line print_prefix item =
```
This code is more complicated that it needs to be: instead of passing `print_prefix` you could just pass the prefix (`!source_tree` or `!build_tree`) and call `print_prefix` with it.

MisterDA replied:
I've rewritten the code, I think it's clearer now. Maybe your comment doesn't apply anymore?

nojb reviewed (commented):
A couple of remarks:

- I find the description of the PR unclear.
  > The new -source-tree option allows to find dependencies in an other location than just the current working directory. The new -build-tree option allows to write targets (build artifacts) in an other location than just the current working directory.

  However, this PR does not "allow" doing this strictly speaking; it only adds a facility to add a (separate) prefix to the targets and the dependencies in its `make` output.

- Could you provide a small example of how this feature would be used in practice in the context of out-of-tree builds? I would like to see a concrete example showing concrete separate source and build trees, a few files in the source tree, the exact invocation of `ocamldep`, the expected output, etc.

shindere commented:
Nicolás Ojeda Bär (2023/05/29 01:38 -0700):
> @nojb commented on this pull request.
> 
> A couple of remarks:
> 
> - I find the description of the PR unclear.
>   > The new -source-tree option allows to find dependencies in an
> other location than just the current working directory. The new
> -build-tree option allows to write targets (build artifacts) in an
> other location than just the current working directory.

That description could certainly be improved, indeed.

I think the purpose of the PR, andinparticular of the `-source-tree`
option, is to be able togenerate rules that willmake sense even if
`make` is not run in the source tree, which will typically be the case
for an out-of-source build where `make` will be run with the top of the
build tree as its current working directory.

> - Could you provide a small example of how this feature would be used
> - in practice in the context of out-of-tree builds? I would like to
> - see a concrete example showing concrete separate source and build
> - trees, a few files in the source tree, the exact invocation of
> - `ocamldep`, the expected output, etc.

@nojb If that would be provided as a test, would that be okay?

I also thought the test would have the double purpose of (1) making sure
the feature works but also (2) demonstrate it, be an example of it.

Re:the computation of the path separator, don't we have this logic
elsewhere already? Can't we just re-use it?

nojb commented:
> I think the purpose of the PR, andinparticular of the `-source-tree` option, is to be able togenerate rules that willmake sense even if `make` is not run in the source tree, which will typically be the case for an out-of-source build where `make` will be run with the top of the build tree as its current working directory.

Thanks. But I'm not sure I understand: if you have your sources in `foo/` and you are building in `foo/build/` and you call `ocamldep` from inside `foo/build/` with `-I ..` then the output of `ocamldep` will contain the relative path of the build directory with respect to the source directory, namely `..`, which means that you can no longer commit the output (and no use of `-source-tree` is going to help in this case). I can imagine that it may work calling `ocamldep` from the *source* directory in combination with `-build-tree` but I am not sure. Or perhaps I am missing something, which is why I would like to see a concrete example of use and better specification of the PR.

> @nojb If that would be provided as a test, would that be okay?

It's the same to me.

MisterDA committed:
Document forgotten ocamldep options in man page

shindere commented:
Nicolás Ojeda Bär (2023/05/30 02:03 -0700):
> Thanks. But I'm not sure I understand: if you have your sources in
> `foo/` and you are building in `foo/build/` and you call `ocamldep`
> from inside `foo/build/` with `-I ..` then the output of `ocamldep`
> will contain the relative path of the build directory with respect to
> the source directory, namely `..`, which means that you can no longer
> commit the output (and no use of `-source-tree` is going to help in
> this case).

If `ocamldep` is invoked without `-all` then it does indeed not refer to
source files and in this situation it is also my understanding that
there is no use for `-source-tree`.

If `-all` is used, though, then rules output by `ocmaldep` will refer to
source files and, currently, they will contain the paths to the source
files relative to where `ocamldep` is run from. That's were, I think,
`-source-tree` becomes handy but assuming you give it a GNU make
variable as argument rather than a relative or absolute path.

Consider for instance the following modular `Hello, world!` example.
`lib.ml` contains:

```
let message = "Hello, world!"
```

And `main.ml` with:

```
let _ = print_endline Lib.message
```

Then, if you are at the top of the source tree and youo run

```
./ocamlc.opt -depend -all -source-tree '$(srcdir)' -build-tree '$(OD)'
lib.ml main.ml
```

you get the following dependencies:

```
$(OD)/lib.cmo $(OD)/lib.cmi : \
    $(srcdir)/lib.ml
$(OD)/lib.cmx $(OD)/lib.o $(OD)/lib.cmi : \
    $(srcdir)/lib.ml
$(OD)/main.cmo $(OD)/main.cmi : \
    $(OD)/lib.cmi \
    $(srcdir)/main.ml
$(OD)/main.cmx $(OD)/main.o $(OD)/main.cmi : \
    $(OD)/lib.cmi \
    $(OD)/lib.cmx \
    $(srcdir)/main.ml
```

Such dependencies can then be stored and if later one runs the build in
a completely different directory, provided that `srcdir` is correctly
defined, it will work.

> I can imagine that it may work calling `ocamldep` from the
> *source* directory in combination with `-build-tree` but I am not
> sure. Or perhaps I am missing something, which is why I would like to
> see a concrete example of use and better specification of the PR.

Yes, sorry, I think we collectively did a bad job at explaining things
so far and I hovpe the above is a step in the direction of clarity.

I shall say that, in my understanding, for `-source-tree` to work
correctly, it will be necessary to change to `$(srcdir)` before running
`ocamldep -all` so that the paths are correct and contain both `srcdir`
and the paths to the subdirs inside the source tree.

I believe as long as `-all` is ot used then neither `cd $(srcdir)`
before calling `ocamldep` nor the `-source-tree` option is necessary.

gasche commented:
(This is my first time learning about the `-all` option of `ocamldep`, which apparently was added by Fabrice Le Fessant in f48f2f229c67661b1b02d4c9df3a5a1feba9cc3b (2012, 4.00).)

shindere commented:
Gabriel Scherer (2023/06/16 01:19 -0700):
> (This is my first time learning about the `-all` option of `ocamldep`,
> which apparently was added by Fabrice Le Fessant in
> f48f2f229c67661b1b02d4c9df3a5a1feba9cc3b.)

Oh! Thanks for the reference, @gasche. I have to say I didn't know about
it either before this work.

MisterDA committed:
Add a -source-tree option to ocamldep

If ocamldep is invoked without `-all`, then the output doesn't refer
to source files and there is no need for `-source-tree`.

If `-all` is used, though, then rules output by ocamldep will refer to
source files and, currently, they will contain the paths to the source
files relative to where ocamldep is run from.

Defining `-source-tree` allows to run the build from a different
directory than of which ocamldep was run.

MisterDA committed:
Test ocamldep new -build-tree and -source-tree options

MisterDA committed:
ocamltest: fix expansion of variables containing escaped $

ocamltest uses Buffer.add_substitute to perform variable substitution.
The function also replaces escaped `$` characters (i.e., the string
`\$`) with the character `$`.
ocamltest would apply substitution in a loop until the input and
substituted strings were equal, however the removal of the escape
character during the first iteration would create variables during the
second iteration.
Consider now that a fixpoint is reached when no substitutions have
been performed.

gasche commented:
I still find this PR confusing: the usage example is provided in the context of using a Makefile, but the feature makes sense when using `-all`, and that option is designed to be used with non-make build systems?

shindere commented:
Gabriel Scherer (2023/06/19 03:12 -0700):
> I still find this PR confusing: the usage example is provided in the
> context of using a Makefile, but the feature makes sense when using
> `-all`, and that option is designed to be used with non-make build
> systems?

I think `-all` could be used in a `Makefile` based build system if, say,
we wouldn't want to rely on pattern-rules.

MisterDA commented:
Let me apologize for the poor quality of the first submission, I think I've improved the code and fixed the problems that were raised. Thanks for the reviews!
We've even found a bug in ocamltest processing of variables.

I second Seb's answer about `-all` and `-source-tree`, this is what I had in mind too but Seb was faster to answer.
I also want to point out that builds systems such as CMake or Meson expect the user to `cd` into the build directory. If we follow that road too, or enable such workflows, we might need to allow ocamldep outputted rules to point back to the source directory when `-all` is used.

nojb commented:
In order to review this PR, it would be useful to have an explicit example of how it is supposed to be used: a small `Makefile`-based project with a few modules and an out-of-tree build and rules for computing the dependencies using the flags introduced here. Otherwise it is difficult to know if the flags proposed here can be used to solve the stated problem.

MisterDA commented:
In light of the discussion here, we've decided to withdraw this PR temporarily until we can re-submit it as part as the larger out-of-source builds effort. Thanks for the reviews so far! We will return…
